packages/zod/src/v4/locales/he.ts

```diff
@@ -67,51 +104,138 @@ const error: () => errors.$ZodErrorMap = () => {
-        if (sizing) {
-          return `קטן מדי: ${issue.origin} צריך להיות ${adj}${issue.minimum.toString()} ${sizing.unit}`;
+        const subject = withDefinite(issue.origin ?? "value");
+
+        if (issue.origin === "string") {
+          // Special handling for strings - more natural Hebrew
+          return `${sizing?.shortLabel ?? "קצר"} מדי: ${subject} צריכה להכיל ${issue.minimum.toString()} ${sizing?.unit ?? ""} ${issue.inclusive ? "או יותר" : "לפחות"}`.trim();
+        }
+
+        if (issue.origin === "number") {
+          // Natural Hebrew for numbers
+          const comparison = issue.inclusive ? `גדול או שווה ל-${issue.minimum}` : `גדול מ-${issue.minimum}`;
+          return `קטן מדי: ${subject} צריך להיות ${comparison}`;
         }
 
-        return `קטן מדי: ${issue.origin} צריך להיות ${adj}${issue.minimum.toString()}`;
+        if (issue.origin === "array" || issue.origin === "set") {
+          // Natural Hebrew for arrays and sets
+          const verb = issue.origin === "set" ? "צריכה" : "צריך";
+          const comparison = issue.inclusive
+            ? `${issue.minimum} ${sizing?.unit ?? ""} או יותר`
+            : `יותר מ-${issue.minimum} ${sizing?.unit ?? ""}`;
+          return `קטן מדי: ${subject} ${verb} להכיל ${comparison}`.trim();
+        }
+
+        const adj = issue.inclusive ? ">=" : ">";
+        const be = verbFor(issue.origin ?? "value");
+        if (sizing?.unit) {
+          return `${sizing.shortLabel} מדי: ${subject} ${be} ${adj}${issue.minimum.toString()} ${sizing.unit}`;
+        }
+        return `${sizing?.shortLabel ?? "קטן"} מדי: ${subject} ${be} ${adj}${issue.minimum.toString()}`;
       }
+
       case "invalid_format": {
         const _issue = issue as errors.$ZodStringFormatIssues;
-        if (_issue.format === "starts_with") return `מחרוזת לא תקינה: חייבת להתחיל ב"${_issue.prefix}"`;
-        if (_issue.format === "ends_with") return `מחרוזת לא תקינה: חייבת להסתיים ב "${_issue.suffix}"`;
-        if (_issue.format === "includes") return `מחרוזת לא תקינה: חייבת לכלול "${_issue.includes}"`;
-        if (_issue.format === "regex") return `מחרוזת לא תקינה: חייבת להתאים לתבנית ${_issue.pattern}`;
-        return `${Nouns[_issue.format] ?? issue.format} לא תקין`;
+        // These apply to strings — use feminine grammar + ה׳ הידיעה
+        if (_issue.format === "starts_with") return `המחרוזת חייבת להתחיל ב "${_issue.prefix}"`;
+        if (_issue.format === "ends_with") return `המחרוזת חייבת להסתיים ב "${_issue.suffix}"`;
+        if (_issue.format === "includes") return `המחרוזת חייבת לכלול "${_issue.includes}"`;
+        if (_issue.format === "regex") return `המחרוזת חייבת להתאים לתבנית ${_issue.pattern}`;
+
+        // Handle gender agreement for formats
+        const noun = Nouns[_issue.format] ?? _issue.format;
+        if (noun === "כתובת אימייל" || noun === "כתובת רשת") {
+          return `${noun} לא תקינה`;
+        }
+        return `${noun} לא תקין`;
```

Comment:
this condition is very specific. i would add gender to the `Nouns` object instead